File: apps/api/echoroo/schemas/annotation_set.py
"""Pydantic schemas for the ground-truth annotation feature (spec 003-annotation).

Covers ``AnnotationSet``, ``AnnotationSegment`` and ``TimeRangeAnnotation``
plus the species-palette and note-attachment payloads.

Schema field naming follows the public OpenAPI contract
(``specs/003-annotation/contracts/``): the wire format exposes
``species_id``, while the ORM layer stores the same value in a column
named ``taxon_id`` because Echoroo's canonical taxonomic entity is
:class:`echoroo.models.taxon.Taxon`. Schemas that hydrate directly from
ORM instances use Pydantic ``AliasChoices`` so both names are accepted
on input and the ORM attribute is picked up correctly on output.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import (
    AliasChoices,
    BaseModel,
    ConfigDict,
    Field,
    model_validator,
)

# ---------------------------------------------------------------------------
# Shared sub-schemas (filters)
# ---------------------------------------------------------------------------


class DateRangeFilter(BaseModel):
    """Inclusive date range ``{"start": "YYYY-MM-DD", "end": "YYYY-MM-DD"}``.

    Persisted as JSONB on ``annotation_sets.filter_date_range``.
    """

    start: str = Field(..., description="Inclusive start date, ISO 8601 (YYYY-MM-DD)")
    end: str = Field(..., description="Inclusive end date, ISO 8601 (YYYY-MM-DD)")


class TimeOfDayRangeFilter(BaseModel):
    """Local time-of-day range; may wrap midnight.

    Persisted as JSONB on ``annotation_sets.filter_time_of_day_range``.
    """

    start: str = Field(..., description="Start time, HH:MM (24h)")
    end: str = Field(..., description="End time, HH:MM (24h)")


# ---------------------------------------------------------------------------
# AnnotationSet
# ---------------------------------------------------------------------------


class AnnotationSetCreate(BaseModel):
    """Request schema for creating an ``AnnotationSet``.

    The background sampling job is dispatched via ``POST
    /annotation-sets/{id}/sample`` after creation; on successful create the
    row is returned with ``status = sampling``.
    """

    project_id: UUID = Field(..., description="Owning project ID")
    dataset_id: UUID = Field(..., description="Source dataset ID")
    name: str = Field(..., min_length=1, max_length=200, description="Display name")
    filter_date_range: DateRangeFilter | None = Field(
        default=None,
        description="Optional inclusive date filter on recording.recorded_at",
    )
    filter_time_of_day_range: TimeOfDayRangeFilter | None = Field(
        default=None,
        description="Optional local time-of-day filter (may wrap midnight)",
    )
    segment_length_sec: int = Field(
        ...,
        ge=10,
        description="Length of every sampled segment in seconds (minimum 10)",
    )
    num_segments: int = Field(
        ...,
        ge=1,
        description="Target number of segments to materialize",
    )


class AnnotationSetUpdate(BaseModel):
    """Request schema for updating an ``AnnotationSet``.

    Sampling parameters cannot be changed once sampling is in progress
    (enforced at the service layer); ``name`` is always mutable.
    """

    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=200,
        description="New display name",
    )
    filter_date_range: DateRangeFilter | None = Field(
        default=None,
        description="New date filter (only accepted before sampling completes)",
    )
    filter_time_of_day_range: TimeOfDayRangeFilter | None = Field(
        default=None,
        description="New time-of-day filter (only accepted before sampling completes)",
    )
    segment_length_sec: int | None = Field(
        default=None,
        ge=10,
        description="New segment length (only accepted before sampling completes)",
    )
    num_segments: int | None = Field(
        default=None,
        ge=1,
        description="New target count (only accepted before sampling completes)",
    )


class AnnotationSetResponse(BaseModel):
    """Response schema for an ``AnnotationSet`` row."""

    id: UUID
    project_id: UUID
    dataset_id: UUID
    created_by_id: UUID
    name: str
    filter_date_range: DateRangeFilter | None = None
    filter_time_of_day_range: TimeOfDayRangeFilter | None = None
    segment_length_sec: int
    num_segments: int
    status: Literal["sampling", "ready", "in_progress", "completed"]
    sampling_warning: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnnotationSetProgress(BaseModel):
    """Per-status segment counts aggregated for an AnnotationSet."""

    total: int
    unannotated: int
    annotated: int
    skipped: int
    empty: int


class PaletteEntryResponse(BaseModel):
    """Response schema for a palette entry.

    ``species_id`` is the wire name for the underlying ``taxon_id`` column.
    """

    species_id: UUID = Field(
        ...,
        validation_alias=AliasChoices("species_id", "taxon_id"),
        serialization_alias="species_id",
    )
    scientific_name: str | None = None
    common_name: str | None = None
    position: int

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class AnnotationSetDetailResponse(AnnotationSetResponse):
    """Detail response with palette and per-status progress counts."""

    palette: list[PaletteEntryResponse] = Field(default_factory=list)
    progress: AnnotationSetProgress


class AnnotationSetListResponse(BaseModel):
    """Paginated list of annotation sets."""

    items: list[AnnotationSetResponse]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# Species palette
# ---------------------------------------------------------------------------


class PaletteItemCreate(BaseModel):
    """Request schema for adding a species to the per-set palette.

    Accepts either ``species_id`` (preferred, matches the contract) or
    ``taxon_id`` (legacy alias).
    """

    species_id: UUID = Field(
        ...,
        validation_alias=AliasChoices("species_id", "taxon_id"),
        serialization_alias="species_id",
        description="Taxon (species) ID to add",
    )
    position: int = Field(
        default=0, ge=0, description="Ordering hint for UI / keyboard slots",
    )

    model_config = ConfigDict(populate_by_name=True)


# ---------------------------------------------------------------------------
# AnnotationSegment
# ---------------------------------------------------------------------------


class AnnotationSegmentResponse(BaseModel):
    """Response schema for an ``AnnotationSegment`` row (list view)."""

    id: UUID
    annotation_set_id: UUID
    recording_id: UUID
    recording_filename: str | None = None
    start_time_sec: float
    end_time_sec: float
    is_empty: bool
    status: Literal["unannotated", "annotated", "skipped"]
    annotated_by_id: UUID | None = None
    annotated_at: datetime | None = None
    annotation_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnnotationSegmentStatusUpdate(BaseModel):
    """Request schema for updating an ``AnnotationSegment``'s lifecycle state.

    - To mark empty: ``is_empty = true`` (service layer rejects this when any
      TimeRangeAnnotation exists).
    - To finalize with existing annotations: ``status = "annotated"``.
    - To skip: ``status = "skipped"``.
    """

    status: Literal["unannotated", "annotated", "skipped"] | None = None
    is_empty: bool | None = None


class AnnotationSegmentListResponse(BaseModel):
    """Paginated list of annotation segments."""

    items: list[AnnotationSegmentResponse]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# TimeRangeAnnotation
# ---------------------------------------------------------------------------


class TimeRangeAnnotationCreate(BaseModel):
    """Request schema for creating a ``TimeRangeAnnotation`` inside a segment.

    Accepts either ``species_id`` or ``taxon_id`` on the wire.
    """

    start_time_sec: float = Field(
        ..., ge=0, description="Start offset in seconds inside the segment",
    )
    end_time_sec: float = Field(
        ..., gt=0, description="End offset in seconds inside the segment",
    )
    species_id: UUID = Field(
        ...,
        validation_alias=AliasChoices("species_id", "taxon_id"),
        serialization_alias="species_id",
        description="Taxon (species) ID",
    )
    confidence: float | None = Field(
        default=None, ge=0.0, le=1.0, description="Optional annotator confidence",
    )

    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="after")
    def _check_range(self) -> TimeRangeAnnotationCreate:
        if self.end_time_sec <= self.start_time_sec:
            raise ValueError("end_time_sec must be greater than start_time_sec")
        return self


class TimeRangeAnnotationUpdate(BaseModel):
    """Request schema for updating a ``TimeRangeAnnotation``.

    All fields are optional; only supplied fields are modified.
    """

    start_time_sec: float | None = Field(default=None, ge=0)
    end_time_sec: float | None = Field(default=None, gt=0)
    species_id: UUID | None = Field(
        default=None,
        validation_alias=AliasChoices("species_id", "taxon_id"),
        serialization_alias="species_id",
    )
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)

    model_config = ConfigDict(populate_by_name=True)


class TimeRangeAnnotationResponse(BaseModel):
    """Response schema for a ``TimeRangeAnnotation`` row."""

    id: UUID
    segment_id: UUID
    start_time_sec: float
    end_time_sec: float
    species_id: UUID = Field(
        ...,
        validation_alias=AliasChoices("species_id", "taxon_id"),
        serialization_alias="species_id",
    )
    species_scientific_name: str | None = None
    species_common_name: str | None = None
    confidence: float | None = None
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime
    note_count: int = 0

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


# ---------------------------------------------------------------------------
# Segment detail (with annotations + notes)
# ---------------------------------------------------------------------------


class AnnotationNoteCreate(BaseModel):
    """Request schema for attaching a note to a segment or TimeRangeAnnotation."""

    content: str = Field(..., min_length=1, max_length=5000)
    is_issue: bool = Field(
        default=False,
        description="Flag this note as raising a quality concern (surfaced in UI)",
    )


class AnnotationNoteResponse(BaseModel):
    """Response schema for a note attached to an annotation entity."""

    id: UUID
    content: str
    is_issue: bool
    is_review: bool
    created_by_id: UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AnnotationSegmentDetailResponse(BaseModel):
    """Response schema for a segment detail view, including children."""

    id: UUID
    annotation_set_id: UUID
    recording_id: UUID
    recording_filename: str | None = None
    recording_duration_sec: float | None = None
    start_time_sec: float
    end_time_sec: float
    is_empty: bool
    status: Literal["unannotated", "annotated", "skipped"]
    annotated_by_id: UUID | None = None
    annotated_at: datetime | None = None
    annotations: list[TimeRangeAnnotationResponse] = Field(default_factory=list)
    notes: list[AnnotationNoteResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Celery dispatch response
# ---------------------------------------------------------------------------


class AnnotationSetSampleDispatchResponse(BaseModel):
    """Response for ``POST /annotation-sets/{id}/sample``."""

    task_id: str
    status: Literal["sampling", "ready", "in_progress", "completed"]

File: apps/web/src/lib/types/annotation-set.ts
/**
 * TypeScript type definitions for the AnnotationSet ground-truth system.
 *
 * Mirrors the OpenAPI contracts under `specs/003-annotation/contracts/`.
 * All species references use the public `species_id` field (the backend
 * maps this to the underlying `taxon_id` column at the schema layer).
 */

// ============================================================
// Enumerations
// ============================================================

/** Lifecycle status of an AnnotationSet. */
export type AnnotationSetStatus =
  | 'sampling'
  | 'ready'
  | 'in_progress'
  | 'completed';

/** Lifecycle status of an individual AnnotationSegment. */
export type AnnotationSegmentStatus =
  | 'unannotated'
  | 'annotated'
  | 'skipped';

/**
 * Segmentation strategy for an AnnotationSet.
 *
 * - `fixed`: fixed-length sliding-window slots (requires `segment_length_sec`).
 * - `whole_recording`: one full-length segment per recording; `num_segments`
 *   then acts as the maximum number of recordings to sample.
 */
export type SegmentMode = 'fixed' | 'whole_recording';

/** Model kind recognised by the cross-model evaluation runner. */
export type EvaluationModelKind = 'birdnet' | 'perch' | 'custom';

/** Status of an evaluation run Celery job. */
export type EvaluationRunStatus =
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed';

// ============================================================
// Filters (shared across create / read)
// ============================================================

/** Inclusive `[start, end]` date filter (ISO-8601 date strings). */
export interface DateRange {
  start: string;
  end: string;
}

/** Inclusive time-of-day filter with `HH:MM` strings. */
export interface TimeOfDayRange {
  start: string;
  end: string;
}

// ============================================================
// Palette
// ============================================================

/** A single species entry in an AnnotationSet's palette. */
export interface PaletteEntry {
  species_id: string;
  scientific_name: string;
  common_name: string | null;
  position: number;
}

/** Request body for adding a palette entry. */
export interface PaletteEntryCreate {
  species_id: string;
  position?: number;
}

// ============================================================
// AnnotationSet
// ============================================================

/** Aggregate progress counts across all segments in a set. */
export interface AnnotationSetProgress {
  total: number;
  unannotated: number;
  annotated: number;
  skipped: number;
  empty: number;
}

/** Summary AnnotationSet record (list view). */
export interface AnnotationSet {
  id: string;
  project_id: string;
  dataset_id: string;
  created_by_id: string;
  name: string;
  filter_date_range: DateRange | null;
  filter_time_of_day_range: TimeOfDayRange | null;
  segment_mode: SegmentMode;
  segment_length_sec: number | null;
  num_segments: number;
  status: AnnotationSetStatus;
  /** Optional human-readable warning surfaced by the sampler. */
  sampling_warning: string | null;
  created_at: string;
  updated_at: string;
  /**
   * Real per-status segment counts for the list view. The list endpoint
   * always populates this; other producers may omit it (null).
   */
  progress: AnnotationSetProgress | null;
}

/** Full AnnotationSet including palette and per-status counts. */
export interface AnnotationSetDetail extends AnnotationSet {
  palette: PaletteEntry[];
  progress: AnnotationSetProgress;
}

/** Request body for `POST /annotation-sets`. */
export interface AnnotationSetCreate {
  project_id: string;
  dataset_id: string;
  name: string;
  filter_date_range?: DateRange | null;
  filter_time_of_day_range?: TimeOfDayRange | null;
  /** Segmentation strategy. Defaults to `fixed` on the backend if omitted. */
  segment_mode?: SegmentMode;
  /** Required when `segment_mode` is `fixed`; omit for `whole_recording`. */
  segment_length_sec?: number | null;
  /**
   * Target segment count. In `whole_recording` mode this is the maximum
   * number of recordings to sample (one full-length segment each).
   */
  num_segments: number;
}

/** Request body for `PATCH /annotation-sets/{id}`. */
export interface AnnotationSetUpdate {
  name?: string;
}

/** Paginated list response wrapper. */
export interface AnnotationSetListResponse {
  items: AnnotationSet[];
  total: number;
  page: number;
  page_size: number;
}

/** Response from dispatching sampling. */
export interface SamplingDispatchResponse {
  task_id: string;
  status: AnnotationSetStatus;
}

// ============================================================
// AnnotationSegment
// ============================================================

/** Summary row returned by the segment listing endpoint. */
export interface AnnotationSegmentSummary {
  id: string;
  recording_id: string;
  recording_filename: string;
  start_time_sec: number;
  end_time_sec: number;
  is_empty: boolean;
  status: AnnotationSegmentStatus;
  annotated_by_id: string | null;
  annotated_at: string | null;
  annotation_count: number;
}

/** Paginated segment list response. */
export interface AnnotationSegmentListResponse {
  items: AnnotationSegmentSummary[];
  total: number;
  page: number;
  page_size: number;
}

/** Request body for `PATCH /segments/{id}`. */
export interface AnnotationSegmentUpdate {
  status?: AnnotationSegmentStatus;
  is_empty?: boolean;
}

/** Query parameters for listing segments. */
export interface ListSegmentsParams {
  status?: AnnotationSegmentStatus;
  is_empty?: boolean;
  page?: number;
  page_size?: number;
}

// ============================================================
// TimeRangeAnnotation
// ============================================================

/** A single time-range annotation attached to a segment. */
export interface TimeRangeAnnotation {
  id: string;
  segment_id: string;
  start_time_sec: number;
  end_time_sec: number;
  species_id: string;
  species_scientific_name: string;
  species_common_name: string | null;
  confidence: number | null;
  created_by_id: string;
  created_at: string;
  updated_at: string;
  note_count: number;
}

/** Request body for creating a TimeRangeAnnotation. */
export interface TimeRangeAnnotationCreate {
  start_time_sec: number;
  end_time_sec: number;
  species_id: string;
  confidence?: number | null;
}

/** Request body for updating a TimeRangeAnnotation. */
export interface TimeRangeAnnotationUpdate {
  start_time_sec?: number;
  end_time_sec?: number;
  species_id?: string;
  confidence?: number | null;
}

// ============================================================
// Notes
// ============================================================

/** A note attached to a segment or annotation. */
export interface AnnotationNote {
  id: string;
  content: string;
  is_issue: boolean;
  created_by_id: string;
  created_at: string;
}

/** Request body for creating a note. */
export interface AnnotationNoteCreate {
  content: string;
  is_issue?: boolean;
}

// ============================================================
// Segment detail (nested annotations + notes)
// ============================================================

/** Full segment detail returned by `GET /segments/{id}`. */
export interface AnnotationSegmentDetail {
  id: string;
  annotation_set_id: string;
  recording_id: string;
  recording_filename: string;
  recording_duration_sec: number;
  start_time_sec: number;
  end_time_sec: number;
  is_empty: boolean;
  status: AnnotationSegmentStatus;
  annotated_by_id: string | null;
  annotated_at: string | null;
  annotations: TimeRangeAnnotation[];
  notes: AnnotationNote[];
}

// ============================================================
// Evaluation
// ============================================================

/** Reference to BirdNET pipeline (no parameters). */
export interface BirdNETModelRef {
  kind: 'birdnet';
}

/** Reference to Perch pipeline (no parameters). */
export interface PerchModelRef {
  kind: 'perch';
}

/** Reference to a specific custom SVM classifier. */
export interface CustomModelRef {
  kind: 'custom';
  model_id: string;
}

/** Discriminated union describing any detection model for evaluation. */
export type EvaluationModelRef = BirdNETModelRef | PerchModelRef | CustomModelRef;

/** Per-species metric row within a model summary. */
export interface SpeciesMetric {
  taxon_id: string;
  scientific_name: string | null;
  common_name: string | null;
  tp_precision: number;
  fp: number;
  tp_recall: number;
  fn: number;
  precision: number;
  recall: number;
  f1: number;
  detections_total: number;
  ground_truths_total: number;
}

/** All-species aggregate metric for a single model reference. */
export interface OverallMetric {
  tp_precision: number;
  fp: number;
  tp_recall: number;
  fn: number;
  precision: number;
  recall: number;
  f1: number;
  detections_total: number;
  ground_truths_total: number;
}

/** Summary bundle for one model reference within an evaluation run. */
export interface ModelEvaluationSummary {
  model_ref: EvaluationModelRef;
  overall: OverallMetric;
  species: SpeciesMetric[];
}

/** Raw evaluation run row (list view / POST response). */
export interface EvaluationRunResponse {
  id: string;
  annotation_set_id: string;
  created_by_id: string;
  status: EvaluationRunStatus;
  requested_model_refs: EvaluationModelRef[];
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

/** Top-level summary response for `GET /evaluation-runs/{id}`. */
export interface EvaluationSummary {
  id: string;
  annotation_set_id: string;
  status: EvaluationRunStatus;
  requested_model_refs: EvaluationModelRef[];
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  models: ModelEvaluationSummary[];
}

/** Paginated list response for evaluation runs. */
export interface EvaluationRunListResponse {
  items: EvaluationRunResponse[];
  total: number;
}

/** Request body for dispatching evaluation. */
export interface EvaluationDispatchRequest {
  model_refs: EvaluationModelRef[];
}
